### PR TITLE
JAMES-2733 Improve RabbitMQ MailQueue getSize performance

### PR DESCRIPTION
**Pull request by chibenwa**
Reopen of https://github.com/linagora/james-project/pull/2332

Choices:

 - Count size by bucket, not by slice. Rationals: this avoids all writes to go at a given time all to the same Cassandra node (scalability issues)
 - Denormalize EnQueueId -> Slice+bucket. Upon dequeue this information needs to be read. Impact expected on performance: low. This needs to be checked. Rationals:
     - Not expose cassandra specific concerns to Rabbit (bucket&slice)
     - For concistency reasons, we publish to rabbit first, then save in the cassandra view. This avoids "concistency issues": if a mail is saved in the view and not published to rabbit, it will never be cleaned up...

### CODE REVIEW

**Comment by chibenwa** [comment]
```
MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:474 » Execution
```

test this please

**Comment by chibenwa** [comment]
test this please (deployment test)

**Comment by Arsnael** [comment]
deployment tests again...

test this please

**Comment by chibenwa** [comment]
```
[494cd9338e8c04f193ff7bb330f27a074db436a9] [ERROR] Failed to execute goal on project apache-james-mailbox-tika: Could not resolve dependencies for project org.apache.james:apache-james-mailbox-tika:jar:3.4.0-SNAPSHOT: Failed to collect dependencies at org.apache.httpcomponents:fluent-hc:jar:4.5.6: Failed to read artifact descriptor for org.apache.httpcomponents:fluent-hc:jar:4.5.6: Could not transfer artifact org.apache.httpcomponents:fluent-hc:pom:4.5.6 from/to central (https://repo.maven.apache.org/maven2): Connection reset -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
:green_apple: before CI reboot

**Comment by chibenwa** [comment]
test this please

**Comment by trantienduchn** [comment]
read it

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java [hunk 1]
```diff
@@ -73,6 +74,26 @@
...
+
+        String QUEUE_NAME = "queueName";
+        String TIME_RANGE_START = "timeRangeStart";
+        String BUCKET_ID = "bucketId";
+
+        String MAIL_COUNT = "mailCount";
+    }
+
+    interface EnqueueIdToBucketTable {
+        String TABLE_NAME = "mailKeyToBucket";
```
Comment (trantienduchn) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java]: can you rename the table to get it match with the interface name?

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java [hunk 2]
```diff
@@ -120,5 +141,21 @@
...
+        .table(BucketSizeTable.TABLE_NAME)
+        .comment("Holds the count of remaining email per bucket")
+        .statement(statement -> statement
+            .addPartitionKey(BucketSizeTable.QUEUE_NAME, text())
+            .addPartitionKey(BucketSizeTable.TIME_RANGE_START, timestamp())
+            .addPartitionKey(BucketSizeTable.BUCKET_ID, cint())
+            .addColumn(BucketSizeTable.MAIL_COUNT, counter()))
+
+        .table(EnqueueIdToBucketTable.TABLE_NAME)
+        .comment("Given a mailKey, resolve the bucket it belongs to")
```
Comment (trantienduchn) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java]: s/Given a mailKey/Given an enqueueId

**Comment by chibenwa** [comment]
To answer you:

1. I think experience prooved us "complexity is inherant to a distributed mail queue". dot. @blackheaven also had report of getSize being too slow. Meaning that the simple model we used was too simple to offer an efficient "getSize" implementation.

That's for complexity.

For cost/benefit analyses and impact on the "write path" we need a Gatling test for that - planned in the upcoming work.

3. The beauty here is that the solution is self healing. If one slice-counter is corrupted, it will only be taken into account while it's newer than the "browseStart" counter. Eventually, the browse start counter will be updated to a newer version, making that faulty slice no longer taken in the computation.

I think we can afford eventual consistency here.

Should we plan adding tests for that?

2. We update it in the "storeMail" and "deleteMail" operation. The choice to get each operation having its single class looks like the root of your concern. Not something that could be easily solved, and not specific to that issue. I don't have better code organisation idea. So I propose to keep it as is.

**Comment by mbaechler** [comment]
1.

I definitely don't buy the "we are used to build complex solution, we should keep going" argument. At all.
The fact we accepted building complex things in the past doesn't mean that it was a good idea nor that we should continue doing so.
The rabbit+cassandra mailqueue implementation started more than one year ago and is still not a production because we keep adding fixes to various issues we have with it.
We don't even have yet feedback about production issues so let's not think it's over.

We failed at using the right tool for the job and everything is then more complex than it should.

My opinion is that we should stop accepting to grow that component for the moment. Adding two more projections is not something I think is wise. We would gain some performance at the cost of correctness (because complexity means bugs).

2. I suggested two solutions. Can you please elaborate why you discarded both?

**Pull request closed without merging**